File: lab/javascript/POO/imprimir.js
class Imprimivel {
  imprimir() {
    throw new Error("O método 'imprimir()' deve ser implementado!");
  }
}

class Relatorio extends Imprimivel {
  constructor(nome, data, descricao) {
    super();
    this.nome = nome;
    this.data = data;
    this.descricao = descricao;
  }

  imprimir() {
    console.log(`Nome do relatório: ${this.nome}`);
    console.log(`Data de Criação: ${this.data}`);
    console.log(`Descrição: ${this.descricao}`);
  }
}

class Etiqueta extends Imprimivel {
  constructor(nome, data, descricao) {
    super();
    this.nome = nome;
    this.data = data;
    this.descricao = descricao;
  }

  imprimir() {
    console.log(`Nome da etiqueta: ${this.nome}`);
    console.log(`Data de Criação: ${this.data}`);
    console.log(`Descrição: ${this.descricao}`);
  }
}

const r = new Relatorio("Confidencial 151 - Fenômeno", "31/10/1995", "SEM REGISTRO");
r.imprimir();
console.log();
const e = new Etiqueta("Lembrete de estudo", "31/10/2025", "Estudar lógica");
e.imprimir();

File: lab/python/POO/imprimir.py
from abc import ABC, abstractmethod


class Imprimivel(ABC):
    @abstractmethod
    def imprimir(self):
        raise ValueError("Imprimir precisa ser chamado!")


class Relatorio(Imprimivel):
    def __init__(self, nome, data, descricao) -> None:
        self.nome = nome
        self.data = data
        self.descricao = descricao

    def imprimir(self):
        print(f"Nome do relatório: {self.nome}")
        print(f"Data de Criação: {self.data}")
        print(f"Descrição: {self.descricao}")


class Etiqueta(Imprimivel):
    def __init__(self, nome, data, descricao) -> None:
        self.nome = nome
        self.data = data
        self.descricao = descricao

    def imprimir(self):
        print(f"Nome da etiqueta: {self.nome}")
        print(f"Data de Criação: {self.data}")
        print(f"Descrição: {self.descricao}")


r = Relatorio("Confidencial 151 - Fenômeno", "31/10/1995", "SEM REGISTRO")
r.imprimir()
print()
e = Etiqueta("Lembrete de estudo", "31/10/2025", "Estudar lógica")
e.imprimir()
